collections/collections.go

```diff
@@ -23,6 +23,18 @@ type StorageProvider interface {
 	KVStore(key storetypes.StoreKey) storetypes.KVStore
 }
 
+// Collection is the interface that all collections support. It will eventually
+// include methods for importing/exporting genesis data and schema
+// reflection for clients.
+type Collection interface {
```

Comment:
I think that:
- Do we want this to be private? Do we want consumers to implement their own collection? I think the only room for expansion of a collection might be in indexes types, although I don't see any other useful index types besides: MultiIndex (secondary key which points to multiple primary keys) and UniqueIndex (secondary key points only to one single primary key). So maybe this can be private.
- A collection is always a Map[K, V],  Item is, under the hood, `Map[noKey, V]`, Sequence is `Item[uint64]`, `KeySet` is `Map[K, noValue]`, `IndexedMap` is one `Map[K, V]` and at least one `KeySet[Pair[IK, PK]]` where `IK` is a field of `V` and `PK` is equal to the `Map`'s `K`, maybe we could abstract everything to be `MapDescriptor{Key: KeyDescriptor, Value: ValueDescriptor, Prefix: []byte, HumanName: string}`?

Reply:
I think private is okay to start. They may use Map under the hood, but I think we should have some distinctions at the schema descriptor level. But let's get into schema descriptors in the next PR

Reply:
I made the collection interface private now